src/mod_muc_db_mnesia.erl

```diff
@@ -0,0 +1,182 @@
+            -> {'aborted', _} | {'atomic', _}.
+store_room(_ServerHost, MucHost, RoomName, Opts) ->
+    F = fun() ->
+                mnesia:write(#muc_room{name_host = {RoomName, MucHost},
+                                       opts = Opts})
+        end,
+    Result = mnesia:transaction(F),
+    case Result of
+        {atomic, _} ->
+            ok;
+        _ ->
+            ?ERROR_MSG("issue=store_room_failed room=~ts", [RoomName])
+    end,
+    Result.
+
+
+restore_room(_ServerHost, MucHost, RoomName) ->
+    try mnesia:dirty_read(muc_room, {RoomName, MucHost}) of
+        [#muc_room{opts = Opts}] ->
+            {ok, Opts};
+        [] ->
+            {error, room_not_found};
+        Other ->
+            {error, Other}
+        catch Class:Reason ->
+            ?ERROR_MSG("issue=restore_room_failed room=~ts reason=~p:~p",
+                       [RoomName, Class, Reason]),
+            {error, {Class, Reason}}
+    end.
+
+-spec forget_room(ejabberd:server(), ejabberd:server(), mod_muc:room()) -> 'ok'.
+forget_room(_ServerHost, MucHost, RoomName) ->
+    F = fun() ->
+                mnesia:delete({muc_room, {RoomName, MucHost}})
+        end,
+    Result = mnesia:transaction(F),
+    case Result of
+        {atomic, _} ->
+            ejabberd_hooks:run(forget_room, MucHost, [MucHost, RoomName]),
+            ok;
+        _ ->
+            ?ERROR_MSG("issue=forget_room_failed room=~ts", [RoomName])
+    end,
+    ok.
+
+get_rooms(_Lserver, MucHost) ->
+    Query = [{#muc_room{name_host = {'_', MucHost}, _ = '_'},
+             [],
+             ['$_']}],
+    mnesia:dirty_select(muc_room, Query).
+
+-spec can_use_nick(ejabberd:server(), ejabberd:server(),
+                   ejabberd:jid(), mod_muc:nick()) -> boolean().
+can_use_nick(_ServerHost, MucHost, JID, Nick) ->
+    LUS = jid_to_lower_us(JID),
+    can_use_nick_internal(MucHost, Nick, LUS).
+
+can_use_nick_internal(MucHost, Nick, LUS) ->
+    Query = [{#muc_registered{us_host = '$1', nick = Nick, _ = '_'},
+             [{'==', {element, 2, '$1'}, MucHost}],
+             ['$_']}],
+    try mnesia:dirty_select(muc_registered, Query) of
+        [] ->
+            true;
+        [#muc_registered{us_host = {U, _Host}}] ->
+            U == LUS
+        catch Class:Reason ->
+            ?ERROR_MSG("issue=can_use_nick_failed jid=~ts nick=~ts reason=~p:~p",
+                       [jid:to_binary(LUS), Nick, Class, Reason]),
+            false
+    end.
+
+get_nick(_ServerHost, MucHost, From) ->
+    LUS = jid_to_lower_us(From),
+    try mnesia:dirty_read(muc_registered, {LUS, MucHost}) of
+        [] ->
+            {error, not_registered};
+        [#muc_registered{nick = Nick}] ->
+            {ok, Nick}
+        catch Class:Reason ->
+            ?ERROR_MSG("issue=get_nick_failed jid=~ts reason=~p:~p",
+                       [jid:to_binary(From), Class, Reason]),
+            {error, {Class, Reason}}
+    end.
+
+set_nick(_ServerHost, _MucHost, _From, <<>>) ->
```

Comment:
I think this should be handled on the frontend-module side, else every backend will have to include this exact clause returning `{error, should_not_be_empty}`.